### ISSUE TITLE
calls to jl_exit()

### ISSUE THREAD

**Issue by dmitry-kazakov** [JuliaLang/julia#30758]
jl_init() and other interfacing functions call to jl_exit(), which in turn calls exit() killing the process. This is quite unacceptable for the case when Julia is embedded into an application written in other language. Could you please replace that with a return code. It would also be nice not to touch the stderr and stdout and use an error string instead.

**Comment by dmitry-kazakov** [comment]
The example is using Julia as a scripting language in an application. Usually Python plays this role. I was considering Julia instead, because Python implementations are unstable.

The scenario is that the user/operator selects in the application's GUI the scripting language he wants to use. E.g. Julia. Then he must specify where Juila is located on his system, e.g. the path to the relocatable library of and the script file he wants to run once or periodically.

Then the application will try to load the Julia's DLL (or *.so) and call to jl_init__threading() from there. And, surprise, the whole process gets killed just because Juilia initialization cannot find something.

The minimal example (in Ada)

```
Load ("D:\Julia-1.0.3\bin\libjulia.dll"); -- Load and bind the library
Init_With_Image ("go to hell");  -- The argument should have been "D:\Julia-1.0.3\bin"
```

The profile of jl_init_with_image etc should rather be:

`   int jl_init_with_image (const char * library, const char * image_path);`

On errors it should do a cleanup and return an error code instead of dropping an atomic bomb on the process.

Regarding error reporting to stderr. There is no stderr in GUI applications. Moreover applications may have reporting and error logging systems of their own. If anything it must be reported there e.g. using a callback function passed to jl_init_xxx (or set by other means):

```
typedef void (*on_error)(const char * error_text, int error_code);
int jl_init_with_image (const char * library, const char * image_path, on_error handler);
```

The default error handler can continue writing onto stderr. That is no problem.

**Comment by yuyichao** [comment]
> `int jl_init_with_image (const char * library, const char * image_path);`

This is impossible.

**Comment by JeffBezanson** [comment]
Setting an error handling procedure is a very tall order, but it seems to me we could have `jl_init` return an error code instead of calling `exit` --- why is it impossible?

**Comment by yuyichao** [comment]
Removing `jl_init` is totally fine. I wasn't replying to that. I was replying to the request to change the signature of `jl_init_with_image` to run the function before loading the library.

**Comment by dmitry-kazakov** [comment]
It is a misunderstanding then. I didn't suggest that jl_init_with_image were called before loading the library, only that it had a return code. Of course a more careful design would deploy an environment object instead of using global variables, as other libraries do. But that is another can of worms.

**Comment by dmitry-kazakov** [comment]
Yes, the relative path to the image can be deduced from the first argument. In the binding to Julia library I am writing it is done, e.g. under Windows the second argument is defaulted to "..\lib\julia\sys.dll", under Debian, Fedora it is to "julia/sys.so".
